### PR TITLE
[CPDLP-378] - Start updating payment breakdowns for npq

### PR DESCRIPTION
**Pull request by StupidCodeFactory**
## Ticket and context

Ticket: https://dfedigital.atlassian.net/browse/CPDLP-378

## Tech review

After quite some time trying to figure out the the dance between `app/services/finance/*` and `lib/payment_calculation/*`,
I've made the decision to decouple the `Finance::NPQ` and `Finance::ECF` by not subclassing the _npq calculator/aggregator_ and _ecf calculator/aggregator_ as they domains are somewhat different and it led to a very convoluted code, quite hard to follow. It took me a while to figure it out.

I've taken the freedom to refactor a couple `FactoryBot` factories:  
1. [`participant_profiles.rb`](https://github.com/DFE-Digital/early-careers-framework/pull/1370/files#diff-26fb0be0ff8b98867d95ce8769b24a3820167e135aca193242bc65c4261dcf57L5-R13). The setup was very confusing and could be done in the way suggested by the `FactoryBot` docs. Ditched the traits in favour or actual factory. This led to a significant amount of `s/create(:participant_profile, :etc)/create(:etc_participant_profile)/`, `s/create(:participant_profile, :mentor)/create(:mentor_participant_profile)/` and `s/create(:participant_profile, :npq)/create(:npq_participant_profile)/`
 However I'm happy to to do this in a separate PR, which we'd merge prior to this. that would certainly make the diff smaller, but the nature of the changes being very repetitive. It didn't seem necessary.
 
2. this made easier to map `:participant_profile` factories to their proper schedule relation, which I needed to create realistic test and seed data. 

### Code quality checks
- [ ] All commit messages are meaningful and true
- [ ] Added enough automated tests

### HTML Checks
- [ ] All new pages have automated accessibility checks
- [ ] All new pages have visual tests via Percy

### Gotchas
- [ ] All `School` queries are correctly scoped - `eligible` when they need to be

### How can someone see it it review app?
1. Click the link to review app (posted by a `github-actions` bot below)
2. Follow the steps from the ticket.

## External API changes

Does this change anything in our external APIs? If so, did you remember to update the CHANGELOG for Lead Providers? (docs/source/release-notes)

### CODE REVIEW

File: spec/components/admin/participants/table_row_spec.rb [hunk 1]
```diff
@@ -12,21 +12,9 @@
...
-    let(:participant_profile) { create :participant_profile, school: nil }
-
-    it { is_expected.to have_css(".govuk-table__cell:nth-child(3)", text: "\n    School\n    \n") }
-  end
-
-  context "when profile is not associated with the school but has school_urn" do
-    let(:participant_profile) { create :participant_profile, school: nil, school_urn: "123456" }
-
-    it { is_expected.to have_content("123456") }
-  end
```
Comment (StupidCodeFactory) [inline spec/components/admin/participants/table_row_spec.rb:31]: Notes: I think this tests weren't very useful as this case can not really happen. can't have a `ParticipantProfile(::ECF)` without a school

File: config/locales/en.yml [hunk 2]
```diff
@@ -111,30 +111,102 @@ en:
...
-              <<: *defaults
-            estimated_teacher_count:
-              <<: *defaults
-              blank: "enter your expected number of teachers"
-        npq_application:
-          attributes:
-            lead_provider_approval_status:
-              has_another_accepted_application: The participant already has an accepted application for this course
-              has_already_been_accepted: This NPQ application has already been accepted
+   attributes:
```
Comment (asmega) [inline config/locales/en.yml:114]: the indentation here is not quite right

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -43,10 +43,11 @@ jobs:
 
     services:
       postgres:
-        image: postgres:11.6-alpine
```
Comment (jamesrose) [inline .github/workflows/test.yml:46]: Can you take these out? They should be reviewed and tested seperately.

Comment (StupidCodeFactory) [reply .github/workflows/test.yml:46]: yes that was a test, the migrations weren't running properly

File: app/helpers/finance_helper.rb [hunk 4]
```diff
@@ -9,6 +23,19 @@ def float_to_percentage(number)
...
 
+  def aggregated_payment(breakdown_array)
+    breakdown_array.map { |breakdown|
+      breakdown[:service_fees][:monthly] + breakdown[:output_payments][:subtotal]
+    }.inject(&:+)
+  end
+
+  # TODO: The VAT flag is set on the ECF lead_provider and not on the cpd_lead_provider table. Once someone is VAT
+  # registered, then it shouldn't matter if they're one, the other or both. VAT should apply evenly. Needs to be
+  # moved and migrated from LeadProvider to CPDLeadProvider.
```
Comment (StupidCodeFactory) [inline app/helpers/finance_helper.rb:34]: That still needs to be addressed, I've asked in the cpd lead provider channel, probably needs to be addressed in a separate PR. Granted we want to copy what was done for ECF payment breakdowns

File: spec/models/participant_declaration/npq_spec.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    end
+
+    it "returns all records when not scoped" do
+      expect(described_class.all.count).to eq(20)
+    end
+
+    it "can retrieve by npq courses attached" do
+      expect(NPQCourse.identifiers.uniq.map { |course_identifier| described_class.for_course(course_identifier).count }.inject(:+)).to eq(20)
+    end
+  end
```
Comment (StupidCodeFactory) [inline spec/models/participant_declaration/npq_spec.rb:18]: I'm not sure this test has a lot of value, it's tested at a much higher level through the feature specs

File: app/models/participant_profile/ecf.rb [hunk 6]
```diff
@@ -39,3 +39,6 @@ def fundable?
     ecf_participant_eligibility&.eligible_status?
   end
 end
+
+require_dependency "participant_profile/ect"
```
Comment (jamesrose) [inline app/models/participant_profile/ecf.rb:43]: nit: should be at the top of the file

File: app/components/finance/npq/contract.html.erb [hunk 7]
```diff
@@ -0,0 +1,69 @@
+<dl class="govuk-summary-list govuk-summary-list--no-border">
+  <div class="govuk-summary-list__row">
+    <dt class="govuk-summary-list__key">
+      <%=t("finance.provider") %>
```
Comment (jamesrose) [inline app/components/finance/npq/contract.html.erb:4]: nit: variable spacing `<%= t(` vs `<%=t` — there are quite a few in this file

File: app/helpers/finance_helper.rb [hunk 8]
```diff
@@ -9,6 +23,19 @@ def float_to_percentage(number)
     number_to_percentage(number * 100, precision: 0)
   end
 
+  def aggregated_payment(breakdown_array)
+    breakdown_array.map { |breakdown|
+      breakdown[:service_fees][:monthly] + breakdown[:output_payments][:subtotal]
+    }.inject(&:+)
```
Comment (jamesrose) [inline app/helpers/finance_helper.rb:29]: Can `breakdown_array` be empty here? You may want to default to `0`, otherwise this would return `nil`.

```suggestion
    }.inject(0, &:+)
```

Comment (StupidCodeFactory) [reply app/helpers/finance_helper.rb:15]: Don't think it could be but that does to hurt

File: app/models/participant_profile/ecf.rb (hunk 6 above)
Comment (StupidCodeFactory) [reply app/models/participant_profile/ecf.rb:43]: @jamesrose intermediate class need to be fully loaded before loading its descendant, otherwise intermediate/parent classes won't know they have descendants, it's all explained in the guides:

https://guides.rubyonrails.org/autoloading_and_reloading_constants_classic_mode.html#autoloading-and-sti

File: app/models/finance/schedule/npq_leadership.rb [hunk 9]
```diff
@@ -1,6 +1,12 @@
 # frozen_string_literal: true
 
 class Finance::Schedule::NPQLeadership < Finance::Schedule
+  IDENTIFIERS = %w[
+    npq-leading-teaching
+    npq-leading-behaviour-culture
+    npq-leading-teaching-development
```
Comment (asmega) [inline app/models/finance/schedule/npq_leadership.rb:7]: these are wrong, see https://github.com/DFE-Digital/early-careers-framework/pull/1467

File: app/models/finance/schedule/npq_specialist.rb [hunk 10]
```diff
@@ -1,6 +1,12 @@
 # frozen_string_literal: true
 
 class Finance::Schedule::NPQSpecialist < Finance::Schedule
+  IDENTIFIERS = %w[
+    npq-senior-leadership
+    npq-headship
+    npq-executive-leadership
```
Comment (asmega) [inline app/models/finance/schedule/npq_specialist.rb:7]: these are wrong, see https://github.com/DFE-Digital/early-careers-framework/pull/1467

File: app/models/participant_declaration/npq.rb [hunk 11]
```diff
@@ -2,4 +2,18 @@
 
 class ParticipantDeclaration::NPQ < ParticipantDeclaration
   include RecordDeclarations::NPQ
+
+  scope :for_course, ->(course_identifier) { where(course_identifier: course_identifier) }
+  scope :for_lead_provider_and_course, ->(cpd_lead_provider, course_identifier) { for_lead_provider(cpd_lead_provider).for_course(course_identifier) }
```
Comment (asmega) [inline app/models/participant_declaration/npq.rb:7]: do we need a compound scope, can't we just chain the building blocks in situ? otherwise we'll have many permutations of compound scopes

File: app/services/finance/npq/participant_eligible_and_payable_aggregator.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+      attr_accessor :cpd_lead_provider, :recorder, :course_identifier
+
+      def initialize(cpd_lead_provider:, course_identifier:, recorder: ParticipantDeclaration::NPQ)
+        self.cpd_lead_provider = cpd_lead_provider
+        self.recorder          = recorder
+        self.course_identifier = course_identifier
+      end
+
+      def aggregate(aggregation_type:, event_type:, interval:)
+        scope = recorder.public_send(self.class.aggregation_types[event_type][aggregation_type], cpd_lead_provider, course_identifier)
```
Comment (asmega) [inline app/services/finance/npq/participant_eligible_and_payable_aggregator.rb:34]: relates to my other comment about compound scopes. this can be done over 2 lines where the `course_identifier` scope can be applied separately

**Review by asmega — APPROVED** [review]
